ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/internal/DarwinUrlUtils.kt

```diff
@@ -5,28 +5,76 @@
-        it.encodedUser = encodedUser?.sanitize(NSCharacterSet.URLUserAllowedCharacterSet)
-        it.encodedPassword = encodedPassword?.sanitize(NSCharacterSet.URLPasswordAllowedCharacterSet)
-
-        val query = encodedQuery.sanitize(NSCharacterSet.URLQueryAllowedCharacterSet)
-
-        it.encodedParameters = ParametersBuilder().apply { appendAll(parseQueryString(query, decode = false)) }
-        it.encodedFragment = encodedFragment.sanitize(NSCharacterSet.URLFragmentAllowedCharacterSet)
-        it.trailingQuery = trailingQuery
-    }.buildString()
-
-    return NSURL(string = urlString)
-}
+    val userEncoded = encodedUser.orEmpty().isEncoded(NSCharacterSet.URLUserAllowedCharacterSet)
+    val passwordEncoded = encodedPassword.orEmpty().isEncoded(NSCharacterSet.URLUserAllowedCharacterSet)
+    val hostEncoded = host.isEncoded(NSCharacterSet.URLHostAllowedCharacterSet)
+    val pathEncoded = encodedPath.isEncoded(NSCharacterSet.URLPathAllowedCharacterSet)
+    val queryEncoded = encodedQuery.isEncoded(NSCharacterSet.URLQueryAllowedCharacterSet)
+    val fragmentEncoded = encodedFragment.isEncoded(NSCharacterSet.URLFragmentAllowedCharacterSet)
+    if (userEncoded && passwordEncoded && hostEncoded && pathEncoded && queryEncoded && fragmentEncoded) {
+        return NSURL(string = toString())
+    }
 
-private fun String.asNSString(): NSString = this as NSString
+    val components = NSURLComponents()
+
+    components.scheme = protocol.name
+
+    components.percentEncodedUser = when {
+        userEncoded -> encodedUser
+        else -> user?.sanitize(NSCharacterSet.URLUserAllowedCharacterSet)
+    }
+    components.percentEncodedPassword = when {
+        passwordEncoded -> encodedPassword
+        else -> password?.sanitize(NSCharacterSet.URLUserAllowedCharacterSet)
+    }
+
+    components.percentEncodedHost = when {
+        hostEncoded -> host
+        else -> host.sanitize(NSCharacterSet.URLHostAllowedCharacterSet)
+    }
+    if (port != DEFAULT_PORT && port != protocol.defaultPort) {
+        components.port = NSNumber(port)
+    }
+
+    components.percentEncodedPath = when {
+        pathEncoded -> encodedPath
+        else -> pathSegments.joinToString("/").sanitize(NSCharacterSet.URLPathAllowedCharacterSet)
+    }
+
+    when {
+        encodedQuery.isEmpty() -> components.percentEncodedQuery = null
+        queryEncoded -> components.percentEncodedQuery = encodedQuery
+        else -> components.percentEncodedQueryItems = parameters.toMap()
+            .flatMap { (key, value) -> if (value.isEmpty()) listOf(key to null) else value.map { key to it } }
+            .map { NSURLQueryItem(it.first.encodeQueryKey(), it.second?.encodeQueryValue()) }
+    }
+
+    components.percentEncodedFragment = when {
+        encodedFragment.isEmpty() -> null
+        fragmentEncoded -> encodedFragment
+        else -> fragment.sanitize(NSCharacterSet.URLFragmentAllowedCharacterSet)
+    }
+
+    return components.URL!!
```

Comment:
we can add something like `?: error("Invalid url: $this") to make it easier for debug